### PR TITLE
shimv2: improve error reporting when the monitor thread fails

### PR DESCRIPTION
**Pull request by jcvenegas**
## Problem
When the monitor thread fails (Vmm or agent  failed to respond). The monitor notifies that something that an error happen and mark the agent as dead. After that, depending on when the notification happen this can result on not well propagated errors (or not event report errors). The main reason is a concurrency problem. 

1. Multiple shim api calls can happen at the same time.
2. The container state is not thread safe, if an error is propagated  and also the container was marked as stopped (after an unexpected error), it can hapend that some api calls like Wait or State just return that the status of the container changed, but the real reason was en error that was not seen because.
3. If an error is well propagated only the most inner error is shown that does not include a lot of information.

## Solution

Improve error reporting:
- virtcontainers: Add errors package
- errors: not modify error if is not grpc type
- errors: use virtcontainers.errors for monitor failing
- monitor: Stop watcher on error

Improve state update of shim container struct
- shim: improve concurrency for container state
- monitor: Stop watcher on error

Example of error trace from kubeadm
```
Normal Scheduled 30s default-scheduler Successfully assigned default/config-env-test-pod to jcvenega-kata-clh-debug Normal Pulling 26s kubelet Pulling image "quay.io/prometheus/busybox:latest" Normal Pulled 26s kubelet Successfully pulled image "quay.io/prometheus/busybox:latest" in 431.185669ms Normal Created 24s kubelet Created container test-container Warning Failed 21s kubelet Error: failed to create containerd task: failed to create shim: Cause: MYERROR

Error trace: 0. MYERROR

Failed to check if Cloud Hypervisor is working
Monitor has failed to watch hypervisor
Monitor marked the agent as dead
Agent was marked dead
ttrpc: closed
Failed to send request grpc.CreateContainerRequest
Agent failed to create container
Failed to create container
Failed to create container in sandbox
shimv2 create failed
The Create() endpoint from kata shim failed
Stack: 0. github.com/kata-containers/kata-containers/src/runtime/virtcontainers/errors.ErrorContext /mnt/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/errors/errors.go +33

github.com/kata-containers/kata-containers/src/runtime/virtcontainers.(*cloudHypervisor).Check /mnt/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/clh.go +791
github.com/kata-containers/kata-containers/src/runtime/virtcontainers.(*monitor).watchHypervisor /mnt/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/monitor.go +153
github.com/kata-containers/kata-containers/src/runtime/virtcontainers.(*monitor).newWatcher.func1 /mnt/go/src/github.com/kata-containers/kata-containers/src/runtime/virtcontainers/monitor.go +64
runtime.goexit /usr/local/go/src/runtime/asm_amd64.s +1581: unknown Warning FailedKillPod 19s (x2 over 20s) kubelet error killing pod: failed to "KillPodSandbox" for "892b6bf8-04e4-477b-92a9-86154e24bcdc" with KillPodSandboxError: "rpc error: code = Unknown desc = failed to stop sandbox container "f0d57bfe72f4d69a5f58f09d4beac291c560a7cbf2467feb7cb199f4367a2df3" in "SANDBOX_READY" state: failed to kill sandbox container: Container.wait\n\tCause: Failed to wait for process from agent\n\tCause: Failed to send request grpc.WaitProcessRequest\n\tCause: ttrpc: closed\n\tCause: Agent was marked dead\n\tCause: Monitor marked the agent as dead\n\tCause: Monitor has failed to watch hypervisor\n\tCause: Failed to check if Cloud Hypervisor is working\n\tCause: MYERROR\n\tCause: Agent was marked dead\n\tCause: Monitor marked the agent as dead\n\tCause: Monitor has failed to watch hypervisor\n\tCause: Failed to check if Cloud Hypervisor is working\n\tCause: MYERROR: unknown"
```

### CODE REVIEW

File: src/runtime/pkg/containerd-shim-v2/container.go
Comment (liubin) [inline src/runtime/pkg/containerd-shim-v2/container.go:44]: Why use a `_` prefix for the fields?

File: src/runtime/pkg/containerd-shim-v2/errors.go
Comment (liubin) [inline src/runtime/pkg/containerd-shim-v2/errors.go:31]: `causeErr` is the golang style.

File: src/runtime/virtcontainers/errors/errors.go
Comment (liubin) [inline src/runtime/virtcontainers/errors/errors.go:1]: New files need a copyright header.

File: src/runtime/virtcontainers/errors/errors.go
Comment (liubin) [inline src/runtime/virtcontainers/errors/errors.go:62]: Suggestion: Replace `return nil` instead of `return err`, though they are the same result, the first one is more simple to understand.

File: src/runtime/virtcontainers/errors/errors.go
Comment (liubin) [inline src/runtime/virtcontainers/errors/errors.go:103]: Blank line can be deleted.

File: src/runtime/virtcontainers/errors/errors.go
Comment (liubin) [inline src/runtime/virtcontainers/errors/errors.go:113]: Can't find `LogErrorNEWERROR` anywhere, should be deleted?

File: src/runtime/virtcontainers/kata_agent.go
Comment (liubin) [inline src/runtime/virtcontainers/kata_agent.go:1990]: This will add 3 lines for each call, I'd suggest wrapping this like:

```golang
addHandler := func(handlerId string, k *kataAgent, f func(ctx context.Context, req interface{}) (interface{}, error)) {
	k.reqHandlers[handlerId] = func(ctx context.Context, req interface{}) (interface{}, error) {
		if err := checkReq(handlerId, k, req); err != nil {
			return nil, err
		}
		return f(ctx, req)
	}
}

addHandler(grpcCheckRequest, k, func(ctx context.Context, req interface{}) (interface{}, error) {
	return k.client.HealthClient.Check(ctx, req.(*grpc.CheckRequest))
})
addHandler(grpcExecProcessRequest, k, func(ctx context.Context, req interface{}) (interface{}, error) {
	return k.client.AgentServiceClient.ExecProcess(ctx, req.(*grpc.ExecProcessRequest))
})
```

File: src/runtime/virtcontainers/kata_agent.go
Comment (liubin) [inline src/runtime/virtcontainers/kata_agent.go:2204]: Two `msgName` checks.

File: src/runtime/pkg/containerd-shim-v2/delete.go
Comment (liubin) [inline src/runtime/pkg/containerd-shim-v2/delete.go:19]: Suggest adding a `TODO` or `FIXME` for this un-normal comment.

File: src/runtime/pkg/containerd-shim-v2/container.go
Comment (jodh-intel) [reply src/runtime/pkg/containerd-shim-v2/container.go:44]: `¯\_(ツ)_/¯`

**Review by jodh-intel — APPROVED** [review]
Thanks @jcvenegas!

lgtm

Could you add a comment showing an example of a current error and an example of how that would look with this new feature?

**Comment by jcvenegas** [comment]
Hey there I got some problems running static checks on github actions, here the error:
```
Error: kata-check.go:16:8: could not import C (cgo preprocessing failed) (typecheck)
import "C"
       ^
make: *** [Makefile:40: static-checks] Error 1
```

I have tried to run golangci-lint locally with typecheck and I did not have any problem, also seems that I am not touching that file. I'll keep looking at it but if you have one hint please let me know.

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

File: src/runtime/pkg/containerd-shim-v2/container.go
Comment (jcvenegas) [reply src/runtime/pkg/containerd-shim-v2/container.go]: I wanted to use `_` prefix because I wanted the struct members where private, and only accessed from the methods, specially because it has to be locked . I have moved to its own package

**Comment by jcvenegas** [comment]
@fidencio  @jodh-intel @liubin  could you give a second review to this PR ?

**Review by jodh-intel — APPROVED** [review]
Thanks @jcvenegas!

lgtm :smile:

**Comment by fidencio** [comment]
Let me close this one in favour of #3685.  Thanks a whole lot, @jcvenegas!

**Pull request closed without merging**
